File: ecbs.cpp
#include "ecbs.h"
#include <iostream>
#include <limits>
#include <algorithm>

ECBS::ECBS(float w_high, float w_low)
    : w_high_(w_high), w_low_(w_low) {}

ECBS::~ECBS() {
    for (auto node : allocated_nodes_) {
        delete node;
    }
}

int ECBS::countAllConflicts(const std::vector<std::vector<Point3D>>& paths, const CollisionDetector& detector) const {
    int num_agents = paths.size();
    int conflicts = 0;
    for (int i = 0; i < num_agents; ++i) {
        for (int j = i + 1; j < num_agents; ++j) {
            if (paths[i].empty() || paths[j].empty()) continue;
            size_t max_size = std::max(paths[i].size(), paths[j].size());
            for (size_t t = 0; t < max_size; ++t) {
                Point3D pi = (t < paths[i].size()) ? paths[i][t] : paths[i].back();
                Point3D pj = (t < paths[j].size()) ? paths[j][t] : paths[j].back();
                if (detector.checkAgentCollision(pi.x, pi.y, pi.z, pj.x, pj.y, pj.z)) {
                    conflicts++;
                }
            }
        }
    }
    return conflicts;
}

bool ECBS::findFirstConflict(const std::vector<std::vector<Point3D>>& paths, const CollisionDetector& detector,
                             int& agent_i, int& agent_j, int& step, Point3D& pi, Point3D& pj) const {
    int num_agents = paths.size();
    for (int i = 0; i < num_agents; ++i) {
        for (int j = i + 1; j < num_agents; ++j) {
            if (paths[i].empty() || paths[j].empty()) continue;
            size_t max_size = std::max(paths[i].size(), paths[j].size());
            for (size_t t = 0; t < max_size; ++t) {
                Point3D pt_i = (t < paths[i].size()) ? paths[i][t] : paths[i].back();
                Point3D pt_j = (t < paths[j].size()) ? paths[j][t] : paths[j].back();
                if (detector.checkAgentCollision(pt_i.x, pt_i.y, pt_i.z, pt_j.x, pt_j.y, pt_j.z)) {
                    agent_i = i;
                    agent_j = j;
                    step = static_cast<int>(t);
                    pi = pt_i;
                    pj = pt_j;
                    return true;
                }
            }
        }
    }
    return false;
}

bool ECBS::plan(const std::vector<Point3D>& starts, const std::vector<Point3D>& goals,
                const VoxelMap& map, const CollisionDetector& detector,
                std::vector<std::vector<Point3D>>& solutions) {
    int M = starts.size();
    if (M == 0) return true;

    // Allocate root node
    CTNode* root = new CTNode(M);
    allocated_nodes_.push_back(root);

    KinodynamicAStar planner(0.5f, 2.0f, 1.0f, w_low_);

    std::cout << "Planning initial paths for " << M << " agents..." << std::endl;
    for (int i = 0; i < M; ++i) {
        std::vector<Point3D> path;
        bool ok = planner.search(starts[i], goals[i], map, detector, {}, {}, path);
        if (!ok) {
            std::cerr << "ECBS Init Error: Failed to find path for agent " << i << std::endl;
            return false;
        }
        root->paths[i] = path;
    }

    // Compute root cost and conflicts
    root->cost = 0;
    for (int i = 0; i < M; ++i) {
        root->cost += root->paths[i].size();
    }
    root->conflicts = countAllConflicts(root->paths, detector);

    std::vector<CTNode*> open_list;
    open_list.push_back(root);

    const int max_ct_nodes = 500;
    int ct_nodes_expanded = 0;

    std::cout << "Starting high-level conflict tree search..." << std::endl;
    while (!open_list.empty() && ct_nodes_expanded++ < max_ct_nodes) {
        // Find minimum cost in OPEN
        float cost_min = std::numeric_limits<float>::max();
        for (const auto node : open_list) {
            if (node->cost < cost_min) {
                cost_min = node->cost;
            }
        }

        // Form FOCAL list: cost <= cost_min * w_high_
        std::vector<size_t> focal_indices;
        float focal_bound = cost_min * w_high_;
        for (size_t i = 0; i < open_list.size(); ++i) {
            if (open_list[i]->cost <= focal_bound) {
                focal_indices.push_back(i);
            }
        }

        // Select the CT node from FOCAL with minimum conflicts, tie-breaker: minimum cost
        size_t best_focal_idx = 0;
        int min_conflicts = std::numeric_limits<int>::max();
        float min_cost = std::numeric_limits<float>::max();

        for (size_t i = 0; i < focal_indices.size(); ++i) {
            size_t open_idx = focal_indices[i];
            CTNode* node = open_list[open_idx];
            if (node->conflicts < min_conflicts) {
                min_conflicts = node->conflicts;
                min_cost = node->cost;
                best_focal_idx = open_idx;
            } else if (node->conflicts == min_conflicts && node->cost < min_cost) {
                min_cost = node->cost;
                best_focal_idx = open_idx;
            }
        }

        // Pop best node
        CTNode* curr = open_list[best_focal_idx];
        open_list.erase(open_list.begin() + best_focal_idx);

        std::cout << "CT Node Expanded: " << ct_nodes_expanded 
                  << " | Cost: " << curr->cost 
                  << " | Conflicts: " << curr->conflicts << std::endl;

        // Check for conflicts
        int agent_i, agent_j, step;
        Point3D pi, pj;
        bool has_conflict = findFirstConflict(curr->paths, detector, agent_i, agent_j, step, pi, pj);

        if (!has_conflict) {
            solutions = curr->paths;
            std::cout << "Conflict-free solutions found! Total CT nodes expanded: " << ct_nodes_expanded << std::endl;
            return true;
        }

        std::cout << "  Conflict found between agent " << agent_i << " and " << agent_j 
                  << " at step " << step << std::endl;

        // Branch 1: Constraint for Agent i
        CTNode* child1 = new CTNode(M);
        child1->paths = curr->paths;
        child1->constraints = curr->constraints;
        
        // Agent i must avoid position pj (agent j's position) at time step
        child1->constraints[agent_i].push_back({ step, pj.x, pj.y, pj.z });
        
        // Re-plan agent i's path
        std::vector<std::vector<Point3D>> other_paths;
        for (int k = 0; k < M; ++k) {
            if (k != agent_i) {
                other_paths.push_back(child1->paths[k]);
            }
        }

        std::vector<Point3D> new_path_i;
        bool ok1 = planner.search(starts[agent_i], goals[agent_i], map, detector, 
                                  child1->constraints[agent_i], other_paths, new_path_i);
        if (ok1) {
            child1->paths[agent_i] = new_path_i;
            child1->cost = 0;
            for (int k = 0; k < M; ++k) child1->cost += child1->paths[k].size();
            child1->conflicts = countAllConflicts(child1->paths, detector);
            
            allocated_nodes_.push_back(child1);
            open_list.push_back(child1);
        } else {
            delete child1;
        }

        // Branch 2: Constraint for Agent j
        CTNode* child2 = new CTNode(M);
        child2->paths = curr->paths;
        child2->constraints = curr->constraints;
        
        // Agent j must avoid position pi (agent i's position) at time step
        child2->constraints[agent_j].push_back({ step, pi.x, pi.y, pi.z });

        // Re-plan agent j's path
        other_paths.clear();
        for (int k = 0; k < M; ++k) {
            if (k != agent_j) {
                other_paths.push_back(child2->paths[k]);
            }
        }

        std::vector<Point3D> new_path_j;
        bool ok2 = planner.search(starts[agent_j], goals[agent_j], map, detector, 
                                  child2->constraints[agent_j], other_paths, new_path_j);
        if (ok2) {
            child2->paths[agent_j] = new_path_j;
            child2->cost = 0;
            for (int k = 0; k < M; ++k) child2->cost += child2->paths[k].size();
            child2->conflicts = countAllConflicts(child2->paths, detector);

            allocated_nodes_.push_back(child2);
            open_list.push_back(child2);
        } else {
            delete child2;
        }
    }

    std::cerr << "Error: ECBS failed to find a conflict-free solution (CT limit reached)." << std::endl;
    return false;
}

File: ecbs.py
import numpy as np
from kinodynamic_astar import KinodynamicAStar

class CTNode:
    def __init__(self, num_agents):
        self.paths = [[] for _ in range(num_agents)]
        self.constraints = [[] for _ in range(num_agents)]
        self.cost = 0.0
        self.conflicts = 0

class ECBS:
    def __init__(self, w_high=1.5, w_low=1.4):
        self.w_high = w_high
        self.w_low = w_low

    def count_all_conflicts(self, paths, detector):
        num_agents = len(paths)
        conflicts = 0
        for i in range(num_agents):
            for j in range(i + 1, num_agents):
                if not paths[i] or not paths[j]:
                    continue
                max_size = max(len(paths[i]), len(paths[j]))
                for t in range(max_size):
                    pi = paths[i][t] if t < len(paths[i]) else paths[i][-1]
                    pj = paths[j][t] if t < len(paths[j]) else paths[j][-1]
                    if detector.check_agent_collision(pi[0], pi[1], pi[2], pj[0], pj[1], pj[2]):
                        conflicts += 1
        return conflicts

    def find_first_conflict(self, paths, detector):
        num_agents = len(paths)
        for i in range(num_agents):
            for j in range(i + 1, num_agents):
                if not paths[i] or not paths[j]:
                    continue
                max_size = max(len(paths[i]), len(paths[j]))
                for t in range(max_size):
                    pi = paths[i][t] if t < len(paths[i]) else paths[i][-1]
                    pj = paths[j][t] if t < len(paths[j]) else paths[j][-1]
                    if detector.check_agent_collision(pi[0], pi[1], pi[2], pj[0], pj[1], pj[2]):
                        return i, j, t, pi, pj
        return None

    def plan(self, starts, goals, voxel_map, detector):
        M = len(starts)
        if M == 0:
            return []

        root = CTNode(M)
        planner = KinodynamicAStar(dt=0.5, v_max=2.0, a_max=1.0, w=self.w_low)

        print(f"ECBS: Planning initial paths for {M} agents...")
        for i in range(M):
            path = planner.search(starts[i], goals[i], voxel_map, detector, [], [])
            if path is None:
                print(f"ECBS: Failed to find initial path for agent {i}")
                return None
            root.paths[i] = path

        root.cost = sum(len(p) for p in root.paths)
        root.conflicts = self.count_all_conflicts(root.paths, detector)

        open_list = [root]
        max_ct_nodes = 500
        ct_nodes_expanded = 0

        while open_list and ct_nodes_expanded < max_ct_nodes:
            ct_nodes_expanded += 1
            
            cost_min = min(node.cost for node in open_list)
            
            focal_bound = cost_min * self.w_high
            focal_nodes = [(node, idx) for idx, node in enumerate(open_list) if node.cost <= focal_bound]
            
            curr, best_idx = min(focal_nodes, key=lambda x: (x[0].conflicts, x[0].cost))
            open_list.pop(best_idx)

            print(f"CT Node Expanded: {ct_nodes_expanded} | Cost: {curr.cost} | Conflicts: {curr.conflicts}")

            conflict = self.find_first_conflict(curr.paths, detector)
            if conflict is None:
                print(f"Conflict-free solution found after expanding {ct_nodes_expanded} nodes.")
                return curr.paths

            agent_i, agent_j, step, pi, pj = conflict
            print(f"  Conflict between agent {agent_i} and {agent_j} at step {step}")

            # Branch 1
            child1 = CTNode(M)
            child1.paths = [list(p) for p in curr.paths]
            child1.constraints = [list(c) for c in curr.constraints]
            child1.constraints[agent_i].append({'step': step, 'x': pj[0], 'y': pj[1], 'z': pj[2]})
            
            other_paths = [child1.paths[k] for k in range(M) if k != agent_i]
            new_path_i = planner.search(starts[agent_i], goals[agent_i], voxel_map, detector, 
                                        child1.constraints[agent_i], other_paths)
            if new_path_i is not None:
                child1.paths[agent_i] = new_path_i
                child1.cost = sum(len(p) for p in child1.paths)
                child1.conflicts = self.count_all_conflicts(child1.paths, detector)
                open_list.append(child1)

            # Branch 2
            child2 = CTNode(M)
            child2.paths = [list(p) for p in curr.paths]
            child2.constraints = [list(c) for c in curr.constraints]
            child2.constraints[agent_j].append({'step': step, 'x': pi[0], 'y': pi[1], 'z': pi[2]})
            
            other_paths = [child2.paths[k] for k in range(M) if k != agent_j]
            new_path_j = planner.search(starts[agent_j], goals[agent_j], voxel_map, detector, 
                                        child2.constraints[agent_j], other_paths)
            if new_path_j is not None:
                child2.paths[agent_j] = new_path_j
                child2.cost = sum(len(p) for p in child2.paths)
                child2.conflicts = self.count_all_conflicts(child2.paths, detector)
                open_list.append(child2)

        print("ECBS: Failed to find conflict-free paths.")
        return None
